test/test_nmt.py

```diff
@@ -1,21 +1,136 @@
 import time
 import unittest
 
+import can
 import canopen
+from canopen.nmt import NMT_STATES, NMT_COMMANDS
 from .util import SAMPLE_EDS
 
 
+class TestNmtBase(unittest.TestCase):
+    def setUp(self):
+        node_id = 2
+        self.node_id = node_id
+        self.nmt = canopen.nmt.NmtBase(node_id)
+
+    def test_send_command(self):
+        dataset = (
+            "OPERATIONAL",
+            "PRE-OPERATIONAL",
+            "SLEEP",
+            "STANDBY",
+            "STOPPED",
+        )
+        for cmd in dataset:
+            with self.subTest(cmd=cmd):
+                code = NMT_COMMANDS[cmd]
+                self.nmt.send_command(code)
+                self.assertNotEqual(self.nmt.state, "INITIALISING")
+
+    def test_state_getset(self):
+        for state in NMT_STATES.values():
+            with self.subTest(state=state):
+                self.nmt.state = state
+                self.assertEqual(self.nmt.state, state)
+
+    def test_state_set_invalid(self):
+        with self.assertRaisesRegex(ValueError, "INVALID"):
+            self.nmt.state = "INVALID"
+
+
+class TestNmtMaster(unittest.TestCase):
+    NODE_ID = 2
+    COB_ID = 0x700 + NODE_ID
+    PERIOD = 0.01
+    TIMEOUT = PERIOD * 2
+
+    def setUp(self):
+        bus = can.ThreadSafeBus(
+            interface="virtual",
+            channel="test",
+            receive_own_messages=True,
+        )
+        net = canopen.Network(bus)
+        net.connect()
+        with self.assertLogs():
+            node = net.add_node(self.NODE_ID, SAMPLE_EDS)
+
+        self.bus = bus
+        self.net = net
+        self.node = node
+
+    def tearDown(self):
+        self.net.disconnect()
+
+    def test_nmt_master_no_heartbeat(self):
+        with self.assertRaisesRegex(canopen.nmt.NmtError, "heartbeat"):
+            self.node.nmt.wait_for_heartbeat(self.TIMEOUT)
+        with self.assertRaisesRegex(canopen.nmt.NmtError, "boot-up"):
+            self.node.nmt.wait_for_bootup(self.TIMEOUT)
+
+    def test_nmt_master_on_heartbeat(self):
+        # Skip the special INITIALISING case.
+        for code in [st for st in NMT_STATES if st != 0]:
+            with self.subTest(code=code):
+                task = self.net.send_periodic(self.COB_ID, [code], self.PERIOD)
+                self.addCleanup(task.stop)
+                actual = self.node.nmt.wait_for_heartbeat(self.TIMEOUT)
+                task.stop()
```

Comment:
Is this safe to be called twice?  You added it as a cleanup callback above.

Reply:
Yeah, it's safe. I added the cleanup task in the case where `wait_for_heartbeat` fails, _but_ the cleanup tasks are not run until the actual test method completes (that is: after the loop ends), so you'll be starting a lot of periodic tasks on the bus, creating completely havoc in the test environment. OTOH, I could just as well use a `try...except NmtError...finally` instead. Let me rewrite this part; I think it will be clearer.

Reply:
I think c2ff667bbb06a4fb2b6a10d5703ce5df211d29f9 may be a little bit easier to read and reason about.